### DISCUSSION TITLE
How to get training log saved to .log file real-time

### DISCUSSION THREAD

**Discussion by cnfuseR** [ultralytics/ultralytics#20558] [Q&A]
I try to get training yolo real-time save to .log file but it looks weird
what i want
![image](https://github.com/user-attachments/assets/a0290b58-4970-47e5-a2d3-57be74435caf)

what I actually got
![image](https://github.com/user-attachments/assets/6049bc08-8012-4c13-bcc0-788447b26949)

**Comment by UltralyticsAssistant** [comment]
👋 Hello @cnfuseR, thank you for reaching out and sharing your question with the Ultralytics community! 🚀 We recommend checking out the [Docs](https://docs.ultralytics.com/) for guidance on logging and training, as well as detailed [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) examples.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us investigate the logging behavior you’re seeing.

If you’re working on custom training ❓, please include as much detail as possible, such as your logging setup, dataset details, and any relevant code or configuration you’re using. Also, review our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) for further insights.

Join the Ultralytics community in your preferred space! For real-time chat, drop by [Discord](https://discord.com/invite/ultralytics) 🎧. For longer discussions, visit [Discourse](https://community.ultralytics.com/). Or share your experience with others on our [Subreddit](https://reddit.com/r/Ultralytics).

## Upgrade

To make sure you are working with the latest features and fixes, upgrade your `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO runs in any of these up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing, verifying correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) across macOS, Windows, and Ubuntu.

This is an automated response to help you get started quickly. An Ultralytics engineer will review your discussion and assist you soon!

**Reply by glenn-jocher** [reply]
Hello @cnfuseR, it looks like your log file contains ANSI color codes that are meant for terminal display but appear as raw escape sequences in your text file. This happens because YOLO's logging is optimized for terminal output.

To get clean real-time logs in a file, you can use the `--no-color` flag with the CLI, or in Python you can redirect stdout after disabling colored output:

```python
from ultralytics import YOLO
from ultralytics.utils.checks import check_requirements
import sys

# Disable color in outputs
from ultralytics.utils.torch_utils import colorstr
import builtins
builtins.colorstr = lambda *args, **kwargs: args[-1]  # override colorstr function

# Redirect output to file
log_file = open('training.log', 'w')
original_stdout = sys.stdout
sys.stdout = log_file

# Train your model
model = YOLO('yolo11n.pt')
model.train(data="coco8.yaml", epochs=3)

# Reset stdout when done
sys.stdout = original_stdout
log_file.close()
```

Alternatively, you can use Python's logging module in a custom callback to write clean logs to a file while training. Let me know if you need a more detailed example of this approach.
